fix(new feat): feat-analyze-missing-skills-in-learning-path

AhmadAlamoush opened the pull request:
Analyze the missing skills if a path could not be computed, then find sub-paths for it if possible.

AhmadAlamoush committed:
fix(new feat): feat-analyze-missing-skills-in-learning-path

Analyze the missing skills if a path could not be computed, then find sub-paths for it if possible.

Elscha commented on src/lib/fastDownward/analysis.ts:
```diff
@@ -0,0 +1,207 @@
+ * Compute which LearningUnits are reversed reachable based on the given goals in state.
+ * Using teachingGoals of the LearningUnits and the skill groups
+ */
+function availableActions<LU extends LearningUnit>(
+	currentState: State,
+	learningUnits: ReadonlyArray<LU>,
+	globalKnowledge: GlobalKnowledge
+) {
+    // Find the LearningUnits for each goal in the state using teachingGoals.
+    let usefulLus = learningUnits.filter(lu => lu.teachingGoals.some(skill => currentState.learnedSkills.includes(skill.id)));
+    
+    // Find the LearningUnits for each goal in the state using globalKnowledge (Skill groups).
+    globalKnowledge.getAllParents().forEach(parent => {
+        if (currentState.getHashCode().includes(parent.id) ) {
+            const lu = learningUnits.filter(lu => lu.teachingGoals.some(skill => parent.nestedSkills.includes(skill.id)));
+            usefulLus = usefulLus.concat(lu);
+        }
+    });
+    
+    // Remove duplicate LearningUnits
+    return [...new Set(usefulLus)];
+}
```
Please refactor:
* fastDownward + analysis contain same function -> Move to separate file (e.g., `utilities.ts`), implement function only once and call them in both algorithms
* Rename analysis to something more meaningful -> For instance, `missinSkillDetection.ts`?

AhmadAlamoush replied:
- Yes, they both have a function with the same name "availableActions" but a different implementation.

- File renames (Solved).

Elscha commented on src/lib/fastDownward/analysis.ts:
```diff
@@ -0,0 +1,207 @@
+    const openList: SearchNode<LU>[] = [new SearchNode<LU>(initialState, null, null, 0, 0)];
+    
+    // Use goalString to find missing skills that does not have a requirement
+    let goalString = `,`.concat(initialState.learnedSkills.join(",").concat(`,`));
```
* Please explain idea of goalString (seems a very importan concept of the algorithm)
* Is a String + many concat operations the best approach or would be an array a better solution?

AhmadAlamoush replied:
- goalString is a string with all the skills for the request goal (or skill) to analyze.

Case 1: If a skill does not have requirements, then we can't find it through skill requirements tracing. 
Case 2: If a skill depends on skill groups (parent/child), then we can't find it through skill requirements tracing.
We use goalString to find the missing skills that do not have requirements and the skill that uses skill groups.

- We use String since it is easy and quick to find and remove skills (better than array). 
- Concat to avoid finding numbers that have a similar number (as string) like "10", is shared between 100, 1005, and 1105. So, we concat commas around it to ",10,".

Elscha commented on src/lib/fastDownward/analysis.ts:
```diff
@@ -0,0 +1,207 @@
+    let goalString = `,`.concat(initialState.learnedSkills.join(",").concat(`,`));
+    
+        while (openList.length > 0) {
+        currentNode = openList.shift()!;
```
Missing indentation. Prettier Addon should atomatically increase the indentation level here.

AhmadAlamoush replied:
Solved

Elscha commented on src/lib/fastDownward/analysis.ts:
```diff
@@ -0,0 +1,207 @@
+ * @param learningUnits All learning units of the system to learn new skills.
+ * @returns A list of the missing skills with the sub paths for them. 
+*/
+export function skillAnalysis<LU extends LearningUnit>(
```
Please add to the comment the idea of your approach. How does it detect missing skills?

AhmadAlamoush replied:
Solved

Elscha commented on src/lib/fastDownward/fastDownward.ts:
```diff
@@ -98,13 +98,37 @@ export function search<LU extends LearningUnit>(
 		const currentNode = openList.shift()!;
 
+		// Get the skill in the goal but not in the currentNode status
+		const remainSkillTemp: string[] = [];
```
Should we add remainSkillTemp to SearchNode / State?

AhmadAlamoush replied:
remainSkillTemp is related to all goal skills (used to check all SearchNode), not related to any specific SearchNode or state.
So, I would suggest keeping it out of the SearchNode / State.

Elscha commented on src/lib/fastDownward/fdFrontend.ts:
```diff
@@ -264,3 +265,30 @@ export function findLearningPath<LU extends LearningUnit>({
 }
+
+/**
+ *
```
Add comments:
* For what is it for
* How does it work

Does this realy belong to fdFrontend (How to call FastDownward) ot should it be moved to PathPlaner (Frntend of the complete library)?

AhmadAlamoush replied:
I tried to follow the same way to call "search" function in fastDownward.ts. 
But that is correct, it shoud be in PathPlaner.
I moved it to PathPlaner (solved).

Elscha commented on src/lib/pathPlanner.ts:
```diff
@@ -50,6 +51,31 @@ export function isAcyclic(
 	return alg.isAcyclic(graph);
 }
 
+/**
```
Add comments:
* For what is it for
* Idea of the approach

AhmadAlamoush replied:
- Add comments: solved

AhmadAlamoush committed:
fix(new feat): feat-analyze-missing-skills-in-learning-path

addressing the requested changes to the skill-lib.

AhmadAlamoush committed:
fix(new feat): feat-analyze-missing-skills-in-learning-path

Add goalString comment.

AhmadAlamoush committed:
fix(new feat): feat-analyze-missing-skills-in-learning-path
Note: Using a string instead of Array or Set in goalString (more research needed)
